Cannot build spicy with Zeek plugin if Homebrew spicy is installed

It looks like as of 61f6147029e510c9ec9be96fdbce593ac48ac90c one cannot compile Spicy with Zeek support from source anymore if one also has Spicy installed from Homebrew.

Roughly doing the following
```
...
brew install spicy --HEAD
cd <SPICY_CHECKOUT>/build && ninja
```
leads to
```
|| ninja: Entering directory `/Users/bbannier/src/spicy/build'
|| [6/25] Building CXX object zeek/plugin/CMakeFiles/Zeek-Spicy.darwin-x86_64.dir/src/protocol-analyzer.cc.o
|| FAILED: zeek/plugin/CMakeFiles/Zeek-Spicy.darwin-x86_64.dir/src/protocol-analyzer.cc.o 
|| /usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++  -DZEEK_PLUGIN_INTERNAL_BUILD=false -DZeek_Spicy_darwin_x86_64_EXPORTS -I../zeek/plugin/src -Izeek/plugin/src -Izeek/plugin -I/usr/local/opt/actor-framework/include -I/usr/local/include -I/usr/local/Cellar/zeek/3.1.2/include/binpac -I/usr/local/Cellar/zeek/3.1.2/include/zeek -I../zeek/plugin/include -I../zeek/compiler/.. -I../spicy/include -Ispicy/include -I../hilti/include -Ihilti/include -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -fPIC   -std=c++17 -MD -MT zeek/plugin/CMakeFiles/Zeek-Spicy.darwin-x86_64.dir/src/protocol-analyzer.cc.o -MF zeek/plugin/CMakeFiles/Zeek-Spicy.darwin-x86_64.dir/src/protocol-analyzer.cc.o.d -o zeek/plugin/CMakeFiles/Zeek-Spicy.darwin-x86_64.dir/src/protocol-analyzer.cc.o -c ../zeek/plugin/src/protocol-analyzer.cc
|| In file included from ../zeek/plugin/src/protocol-analyzer.cc:5:
|| In file included from /usr/local/include/zeek-spicy/runtime-support.h:26:
/usr/local/Cellar/spicy/HEAD-61f6147/include/zeek-spicy/cookie.h|22 col 8| error: redefinition of 'ProtocolAnalyzer'
|| struct ProtocolAnalyzer {
||        ^
zeek/plugin/include/cookie.h|22 col 8| note: previous definition is here
|| struct ProtocolAnalyzer {
||        ^
|| In file included from ../zeek/plugin/src/protocol-analyzer.cc:5:
|| In file included from /usr/local/include/zeek-spicy/runtime-support.h:26:
/usr/local/Cellar/spicy/HEAD-61f6147/include/zeek-spicy/cookie.h|31 col 8| error: redefinition of 'FileAnalyzer'
|| struct FileAnalyzer {
||        ^
zeek/plugin/include/cookie.h|31 col 8| note: previous definition is here
|| struct FileAnalyzer {
||        ^
|| 2 errors generated.
|| [7/25] Building CXX object zeek/plugin/CMakeFiles/Zeek-Spicy.darwin-x86_64.dir/src/file-analyzer.cc.o
|| FAILED: zeek/plugin/CMakeFiles/Zeek-Spicy.darwin-x86_64.dir/src/file-analyzer.cc.o 
|| /usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++  -DZEEK_PLUGIN_INTERNAL_BUILD=false -DZeek_Spicy_darwin_x86_64_EXPORTS -I../zeek/plugin/src -Izeek/plugin/src -Izeek/plugin -I/usr/local/opt/actor-framework/include -I/usr/local/include -I/usr/local/Cellar/zeek/3.1.2/include/binpac -I/usr/local/Cellar/zeek/3.1.2/include/zeek -I../zeek/plugin/include -I../zeek/compiler/.. -I../spicy/include -Ispicy/include -I../hilti/include -Ihilti/include -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -fPIC   -std=c++17 -MD -MT zeek/plugin/CMakeFiles/Zeek-Spicy.darwin-x86_64.dir/src/file-analyzer.cc.o -MF zeek/plugin/CMakeFiles/Zeek-Spicy.darwin-x86_64.dir/src/file-analyzer.cc.o.d -o zeek/plugin/CMakeFiles/Zeek-Spicy.darwin-x86_64.dir/src/file-analyzer.cc.o -c ../zeek/plugin/src/file-analyzer.cc
|| In file included from ../zeek/plugin/src/file-analyzer.cc:5:
|| In file included from /usr/local/include/zeek-spicy/runtime-support.h:26:
/usr/local/Cellar/spicy/HEAD-61f6147/include/zeek-spicy/cookie.h|22 col 8| error: redefinition of 'ProtocolAnalyzer'
|| struct ProtocolAnalyzer {
||        ^
zeek/plugin/include/cookie.h|22 col 8| note: previous definition is here
|| struct ProtocolAnalyzer {
||        ^
|| In file included from ../zeek/plugin/src/file-analyzer.cc:5:
|| In file included from /usr/local/include/zeek-spicy/runtime-support.h:26:
/usr/local/Cellar/spicy/HEAD-61f6147/include/zeek-spicy/cookie.h|31 col 8| error: redefinition of 'FileAnalyzer'
|| struct FileAnalyzer {
||        ^
zeek/plugin/include/cookie.h|31 col 8| note: previous definition is here
|| struct FileAnalyzer {
||        ^
|| 2 errors generated.
|| [17/25] Building CXX object spicy/CMakeFiles/spicy.dir/src/autogen/__parser.cc.o
|| ninja: build stopped: subcommand failed.
```

It seems we run into this since the inherited Zeek CMake files add `/usr/local/include` to the path where we find `/usr/local/include/zeek-spicy`. We then find `ProtocolAnalyzer` from both our checkout and `/usr/include`. The correct way would be to exclusively include any Spicy headers from the checkout.

_EDIT_

I think the issue is due to `/usr/local/include` being in the include lookup path (for Zeek headers), and us finding at least `cookie.h` in both that path and the current checkout from different include directives (I see that `cookie.h` is included differently that e.g., `runtime-support.h` which might be related). In any case, for `#pragma one` these files for `cookie.h` are different files (an include guard would have lead to us including the definitions exactly once).

--- a/3rdparty/justrx/CMakeLists.txt
+++ b/3rdparty/justrx/CMakeLists.txt
@@ -31,13 +31,13 @@ set(SOURCES
 
 add_library(jrx ${SOURCES})
 # target_link_libraries(jrx ${FLEX_LIBRARIES})
-target_include_directories(jrx PUBLIC include)
-target_include_directories(jrx PUBLIC ${CMAKE_CURRENT_BINARY_DIR})
+target_include_directories(jrx BEFORE PUBLIC include)
+target_include_directories(jrx BEFORE PUBLIC ${CMAKE_CURRENT_BINARY_DIR})
 
 add_library(jrx-objects OBJECT ${SOURCES})
 target_compile_options(jrx-objects PRIVATE "-fPIC")
 # target_link_libraries(jrx-objects ${FLEX_LIBRARIES})
-target_include_directories(jrx-objects PUBLIC include)
-target_include_directories(jrx-objects PUBLIC ${CMAKE_CURRENT_BINARY_DIR})
+target_include_directories(jrx-objects BEFORE PUBLIC include)
+target_include_directories(jrx-objects BEFORE PUBLIC ${CMAKE_CURRENT_BINARY_DIR})
 
 add_subdirectory(test)

--- a/3rdparty/justrx/test/CMakeLists.txt
+++ b/3rdparty/justrx/test/CMakeLists.txt
@@ -1,5 +1,4 @@
-
-include_directories(${CMAKE_CURRENT_SOURCE_DIR}/../src)
+include_directories(BEFORE ${CMAKE_CURRENT_SOURCE_DIR}/../src)
 
 add_executable(retest retest.c)
 target_link_libraries(retest jrx)

--- a/cmake/FindClangJIT.cmake
+++ b/cmake/FindClangJIT.cmake
@@ -170,7 +170,7 @@ if ( CLANG_JIT_FOUND )
     message(STATUS "Linking against clang libraries '${clang_libs}'")
 
     add_library(clang-jit INTERFACE)
-    target_include_directories(clang-jit INTERFACE ${LLVM_INCLUDE_DIRS})
+    target_include_directories(clang-jit BEFORE INTERFACE ${LLVM_INCLUDE_DIRS})
     target_compile_definitions(clang-jit INTERFACE ${LLVM_DEFINITIONS})
     target_link_options(clang-jit INTERFACE "LINKER:-rpath;${LLVM_LIBRARY_DIR}")
     target_link_libraries(clang-jit INTERFACE ${llvm_libs} ${clang_libs})

--- a/hilti/CMakeLists.txt
+++ b/hilti/CMakeLists.txt
@@ -122,8 +122,8 @@ target_compile_options(hilti PRIVATE "-Wall")
 target_link_libraries(hilti PRIVATE $<IF:$<CONFIG:Debug>,hilti-rt-debug-objects,hilti-rt-objects>)
 target_link_libraries(hilti PRIVATE $<$<BOOL:${HILTI_HAVE_JIT}>:clang-jit>)
 target_link_libraries(hilti PRIVATE std::filesystem)
-target_include_directories(hilti PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
-target_include_directories(hilti PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>)
+target_include_directories(hilti BEFORE PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
+target_include_directories(hilti BEFORE PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>)
 
 # Unclear why we need this: Without it, the generated Bison/Flex get a broken
 # include path on some systems. (Seen on Ubuntu 19.10).
@@ -164,9 +164,9 @@ foreach ( lib hilti-rt hilti-rt-debug )
     add_dependencies(${lib}-objects version)
     target_compile_options(${lib}-objects PRIVATE "-fPIC")
     target_link_libraries(${lib}-objects PRIVATE std::filesystem ${CMAKE_DL_LIBS})
-    target_include_directories(${lib}-objects PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
-    target_include_directories(${lib}-objects PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>)
-    target_include_directories(${lib}-objects PRIVATE include/3rdparty/libtask)
+    target_include_directories(${lib}-objects BEFORE PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
+    target_include_directories(${lib}-objects BEFORE PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>)
+    target_include_directories(${lib}-objects BEFORE PRIVATE include/3rdparty/libtask)
 
     # The following should instead just be:
     #
@@ -177,7 +177,7 @@ foreach ( lib hilti-rt hilti-rt-debug )
     # referenced target is built before the dependent target is linked, we also
     # explicitly depend on jrx-objects to make sure all its objects are build.
     add_dependencies(${lib}-objects jrx-objects)
-    target_include_directories(${lib}-objects PRIVATE ${CMAKE_SOURCE_DIR}/3rdparty/justrx/include)
+    target_include_directories(${lib}-objects BEFORE PRIVATE ${CMAKE_SOURCE_DIR}/3rdparty/justrx/include)
     target_link_libraries(${lib}-objects      PRIVATE $<TARGET_OBJECTS:jrx-objects>)
 
     add_library(${lib} STATIC)

--- a/spicy/CMakeLists.txt
+++ b/spicy/CMakeLists.txt
@@ -79,8 +79,8 @@ target_compile_options(spicy PRIVATE "-Wall")
 target_link_libraries(spicy  PRIVATE $<IF:$<CONFIG:Debug>,spicy-rt-debug-objects,spicy-rt-objects>)
 target_link_libraries(spicy  PUBLIC hilti)
 target_link_libraries(spicy  PRIVATE std::filesystem)
-target_include_directories(spicy PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
-target_include_directories(spicy PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>)
+target_include_directories(spicy BEFORE PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
+target_include_directories(spicy BEFORE PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>)
 
 # Unclear why we need this: Without it, the generated Bison/Flex get a broken
 # include path on some systems. (Seen on Ubuntu 19.10).
@@ -107,9 +107,9 @@ foreach ( lib spicy-rt spicy-rt-debug )
     target_compile_options(${lib}-objects PRIVATE "-fPIC")
     target_link_libraries(${lib}-objects PRIVATE std::filesystem)
     target_link_libraries(${lib}-objects PUBLIC ZLIB::ZLIB)
-    target_include_directories(${lib}-objects PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
-    target_include_directories(${lib}-objects PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>)
-    target_include_directories(${lib}-objects PRIVATE include/3rdparty/libtask)
+    target_include_directories(${lib}-objects BEFORE PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
+    target_include_directories(${lib}-objects BEFORE PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>)
+    target_include_directories(${lib}-objects BEFORE PRIVATE include/3rdparty/libtask)
 
     add_library(${lib} STATIC)
     target_link_libraries(${lib} ${lib}-objects)

--- a/zeek/CMakeLists.txt
+++ b/zeek/CMakeLists.txt
@@ -6,6 +6,6 @@ add_subdirectory(compiler)
 add_executable(spicyz bin/spicyz.cc)
 target_compile_options(spicyz PRIVATE "-Wall")
 target_link_libraries(spicyz PRIVATE zeek-compiler spicy)
-target_include_directories(spicyz PRIVATE ${CMAKE_CURRENT_BINARY_DIR}/plugin)
+target_include_directories(spicyz BEFORE PRIVATE ${CMAKE_CURRENT_BINARY_DIR}/plugin)
 
 install(TARGETS spicyz RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})

--- a/zeek/compiler/CMakeLists.txt
+++ b/zeek/compiler/CMakeLists.txt
@@ -6,6 +6,6 @@ set(SOURCES
     )
 
 add_library(zeek-compiler OBJECT ${SOURCES})
-target_include_directories(zeek-compiler PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/..")
+target_include_directories(zeek-compiler BEFORE PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/..")
 target_compile_options(zeek-compiler PRIVATE "-fPIC")
 target_link_libraries(zeek-compiler PRIVATE spicy)

--- a/zeek/plugin/CMakeLists.txt
+++ b/zeek/plugin/CMakeLists.txt
@@ -53,11 +53,11 @@ else ()
     set(ZEEK_DEBUG_BUILD_CXX "false")
 endif ()
 
-target_include_directories(${_plugin_lib} PRIVATE include)
+target_include_directories(${_plugin_lib} BEFORE PRIVATE include)
 
 # Without JIT support, we won't have the "compiler/" include directory available
 # through the "zeek-compiler" target, so add it no matter what.
-target_include_directories(${_plugin_lib} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/..)
+target_include_directories(${_plugin_lib} BEFORE PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/..)
 
 # For JIT support.
 target_link_libraries(${_plugin_lib} $<$<BOOL:${ZEEK_HAVE_JIT}>:zeek-compiler>)

